src/transformers/models/grounding_dino/modeling_grounding_dino.py

```diff
@@ -2904,18 +2915,58 @@ def forward(self, outputs, targets):
 
         if "enc_outputs" in outputs:
             enc_outputs = outputs["enc_outputs"]
-            bin_targets = copy.deepcopy(targets)
-            for bt in bin_targets:
-                bt["class_labels"] = torch.zeros_like(bt["class_labels"])
-            indices = self.matcher(enc_outputs, bin_targets)
+            indices = self.matcher(enc_outputs, targets)
             for loss in self.losses:
-                l_dict = self.get_loss(loss, enc_outputs, bin_targets, indices, num_boxes)
+                l_dict = self.get_loss(loss, enc_outputs, targets, indices, num_boxes)
                 l_dict = {k + "_enc": v for k, v in l_dict.items()}
                 losses.update(l_dict)
 
         return losses
 
 
+def build_label_maps(logits, input_ids):
+    """
+    Computes a mapping between the tokens associated with the prompt labels in the logit space with shape `(batch_size, num_labels, hidden_size)`
+    where `num_labels` is defined by the number of classes in the input prompt.
+
+    For instance, given the prompt "fish. shark." we get input_ids = [  101,  3869,  1012, 11420,  1012,   102].
+    This function will return a mapping for each of the prompt tokens (i.e. tokens associated with "fish" and "shark")
+    indicating their position in the logit space.
+
+    """
+    hidden_size = logits.shape[-1]
+    # Add [PAD] token to the list of special tokens
+    delimiter_tokens = torch.tensor(SPECIAL_TOKENS + [0], device=input_ids.device)
+
+    delimiter_token_masks = torch.isin(input_ids, delimiter_tokens)
+    label_maps = ()
+    for delimiter_token_mask in delimiter_token_masks:
+        label_map_within_batch = []
+        delimiter_indices = torch.where(delimiter_token_mask)[0]
+        for i in range(len(delimiter_indices) - 1):
+            start = delimiter_indices[i]
+            end = delimiter_indices[i + 1]
+            if end - start > 1:
+                label_map = torch.zeros(hidden_size, device=input_ids.device)
+                label_map[start + 1 : end] = 1
+                label_map_within_batch.append(label_map)
+
+        label_maps += (torch.stack(label_map_within_batch),)
```

Comment:
This looks like it can be vectorized

Reply:
indeed, just gotta loop over the batch size dim as we can have a different number of labels within a batch